### PR TITLE
Better initializer support in `L.Maxout`

### PR DESCRIPTION
**Pull request by takagi**
Merge #5058 and #5064 first.

The documentation of `L.Maxout` says that `initialW` and `initial_bias` can be initializers, but its implementation actually looks assuming that they are scalars or ndarrays.

This PR fixes `L.Maxout` to support general initializers including `Initializer` instances and callables.

Note that callables are wrapped to destructively reshape arrays to initialize into appropreate shapes for L.Maxout.

### CODE REVIEW

File: chainer/links/activation/maxout.py [hunk 1]
```diff
@@ -59,19 +60,46 @@ def __init__(self, in_size, out_size, pool_size,
...
+        if numpy.isscalar(initialW):
+            pass
+        elif isinstance(initialW, (numpy.ndarray, cuda.ndarray)):
+            if initialW.ndim != 3:
+                raise ValueError('initialW.ndim should be 3')
+            initialW = initialW.reshape(linear_out_size, in_size)
+        elif isinstance(initialW, initializer.Initializer):
+            pass
+        elif callable(initialW):
+            def make_initialW_wrapper(initialW):
```
Comment (beam2d) [inline chainer/links/activation/maxout.py]: I don't think we need such a rich func wrapper here. How about making it simpler as follows:
```py
        initialW_orig = initialW
        def initialW(array):
            array.shape = (out_size, pool_size, in_size)
            initialW_orig(array)
            array.shape = (linear_out_size, in_size)
```

File: chainer/links/activation/maxout.py [hunk 2]
```diff
@@ -59,19 +60,46 @@ def __init__(self, in_size, out_size, pool_size,
...
+        if numpy.isscalar(initial_bias):
+            pass
+        elif isinstance(initial_bias, (numpy.ndarray, cuda.ndarray)):
+            if initial_bias.ndim != 2:
+                raise ValueError('initial_bias.ndim should be 2')
+            initial_bias = initial_bias.reshape(linear_out_size)
+        elif isinstance(initial_bias, initializer.Initializer):
+            pass
+        elif callable(initial_bias):
+            def make_initial_bias_wrapper(initial_bias):
```
Comment (beam2d) [inline chainer/links/activation/maxout.py]: ditto (simpler wrapper)

File: chainer/links/activation/maxout.py [hunk 3]
```diff
@@ -59,19 +60,46 @@ def __init__(self, in_size, out_size, pool_size,
...
+        elif callable(initialW):
+            def make_initialW_wrapper(initialW):
+                @functools.wraps(initialW)
+                def wrapper(array):
+                    array.shape = (out_size, pool_size, in_size)
+                    initialW(array)
+                    array.shape = (linear_out_size, in_size)
+                return wrapper
+            initialW = make_initialW_wrapper(initialW)
+        else:
```
Comment (beam2d) [inline chainer/links/activation/maxout.py]: Remove this clause.

File: chainer/links/activation/maxout.py [hunk 4]
```diff
@@ -59,19 +60,46 @@ def __init__(self, in_size, out_size, pool_size,
...
+        elif callable(initial_bias):
+            def make_initial_bias_wrapper(initial_bias):
+                @functools.wraps(initial_bias)
+                def wrapper(array):
+                    array.shape = (out_size, pool_size)
+                    initial_bias(array)
+                    array.shape = linear_out_size,
+                return wrapper
+            initial_bias = make_initial_bias_wrapper(initial_bias)
+        else:
```
Comment (beam2d) [inline chainer/links/activation/maxout.py]: ditto (remove)

**Comment by okuta** [comment]
Please rebase.

**Comment by takagi** [comment]
I rebased it and will fix it for the comments.

File: chainer/links/activation/maxout.py (hunk 1 above)
Comment (takagi) [reply chainer/links/activation/maxout.py]: oh... fixed!

File: chainer/links/activation/maxout.py (hunk 2 above)
Comment (takagi) [reply chainer/links/activation/maxout.py]: Fixed!

File: chainer/links/activation/maxout.py (hunk 3 above)
Comment (takagi) [reply chainer/links/activation/maxout.py]: Removed!

File: chainer/links/activation/maxout.py (hunk 4 above)
Comment (takagi) [reply chainer/links/activation/maxout.py]: Removed!

**Comment by takagi** [comment]
Jenkins, test this please.

**Comment by takagi** [comment]
Jenkins, test this please.

File: chainer/links/activation/maxout.py [hunk 5]
```diff
@@ -59,19 +59,38 @@ def __init__(self, in_size, out_size, pool_size,
...
-                initial_bias = numpy.full(
-                    (linear_out_size,), initial_bias, dtype=dtype)
-            elif isinstance(initial_bias, (numpy.ndarray, cuda.ndarray)):
-                initial_bias = initial_bias.reshape(linear_out_size)
-            else:
-                raise ValueError(
-                    'initial bias must be float, ndarray, or None')
+        if numpy.isscalar(initialW):
+            pass
+        elif isinstance(initialW, (numpy.ndarray, cuda.ndarray)):
```
Comment (okuta) [inline chainer/links/activation/maxout.py]: Please use `is_arrays_compatible`.

Comment (takagi) [reply chainer/links/activation/maxout.py]: Fixed!

File: chainer/links/activation/maxout.py [hunk 6]
```diff
@@ -59,19 +59,39 @@ def __init__(self, in_size, out_size, pool_size,
...
-            if numpy.isscalar(initial_bias):
-                dtype = chainer.get_dtype()
-                initial_bias = numpy.full(
-                    (linear_out_size,), initial_bias, dtype=dtype)
-            elif isinstance(initial_bias, (numpy.ndarray, cuda.ndarray)):
-                initial_bias = initial_bias.reshape(linear_out_size)
-            else:
-                raise ValueError(
-                    'initial bias must be float, ndarray, or None')
+        if numpy.isscalar(initialW):
```
[the remark this answers is not in the document]
Comment (takagi) [reply chainer/links/activation/maxout.py]: Fixed!

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**